contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,816 @@
+ of the former. 
+ - In place configuration update via SIGHUP is not universally 
+ supported, and Kubelet provides no mechanism to perform this currently. Pod 
+ reconfiguration will be performed destructively.
+ - Stateful applications are likely to evolve wire protocols and storage formats
+  between versions. In most cases, when updating the application's Pod's 
+  containers, it will not be safe to roll back or forward to an arbitrary 
+  version. Controller based Pod update should work well when rolling out an 
+  update, or performing a rollback, between two specific revisions of the 
+  controlled API object. This is how Deployment functions, and this property is,
+  perhaps, even more critical for stateful applications.
+
+## Requirements
+This design is based on the following requirements.
+- Users must be able to update the containers of a StatefulSet's Pods.
+  - Updates to container commands, images, resources and configuration must be 
+  supported.
+- The update must progress in a sequential, deterministic order and respect the 
+  StatefulSet
+  [identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+  [deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+  guarantees.
+- A failed update must halt.
+- Users must be able to rollback an update.
+- Users must be able to roll forward to fix a failing/failed update.
+- Users must be able to view the status of an update.
+- Users should be able to view a bounded history of the updates that have been 
+applied to the StatefulSet.
+
+## API Object
+
+The following modifications will be made to the StatefulSetSpec API object.
+
+```go
+// StatefulSetUpdateStrategy indicates the strategy that the StatefulSet 
+// controller will use to perform updates. It includes any additional parameters 
+// nessecary to preform the update for the indicated strategy.
+type StatefulSetUpdateStrategy struct {
+        // Type indicates the type of the StatefulSetUpdateStrategy.
+        Type StatefulSetUpdateStrategyType
+        // Partition is used to communicate the ordinal at which to partition 
+        // the StatefulSet when Type is PartitionedStatefulSetStrategyType. This 
+        // value must be set when Type is PartitionedStatefulSetStrategyType, 
+        // and it must be nil otherwise.
+        Partition *ParitionedStatefulSet
+}
+
+// StatefulSetUpdateStrategyType is a string enumeration type that enumerates 
+// all possible update strategies for the StatefulSet controller.
+type StatefulSetUpdateStrategyType string
+
+const (
+    // PartitionedStatefulSetStrategyType indicates that updates will only be 
+    // applied to a partition of the StatefulSet. This is useful for canaries 
+    // and phased roll outs.
+    PartitionedStatefulSetStrategyType StatefulSetUpdateStrategyType = "Partitioned"
```

Comment:
To me Partitioned strategy seem like a special case of RollingUpdate, it is just paused in that state. It would make building integrations (controlling the Ordinal) on top of it easier.